### ISSUE TITLE
Feature Request: Enable Disabling Linters via Source

### ISSUE THREAD

**Issue by mherchel** [sasstools/sass-lint#70]
SCSS-lint has the ability to disable the linters within the source files

https://github.com/brigade/scss-lint#disabling-linters-via-source

This is unfortunately important for my project's workflow :)

**Comment by Snugug** [comment]
I'm sorry to hear that :stuck_out_tongue_closed_eyes: 

This isn't a high priority at the moment with a slew of rules that need to be written and some bugs to be squashed, but I'd be happy to take a PR for it. I'd expect the abstraction for determining if a rule should be enabled/disabled to live in [helpers](https://github.com/sasstools/sass-lint/blob/master/lib/helpers.js)

**Comment by cgtm** [comment]
Yeah, this would be hugely useful for the large project I'm working on. We've only recently started linting our scss files (I know!) and we've been using scss-lint, peppering all of our current files with overrides to get them to pass for now (as described by mherchel above).

As we're moving from scss-lint to sass-lint (ditching as much ruby stuff as possible), it would be great to have this functionality. Otherwise I can see needing to create multiple rulesets for different sets of files (being loose with everything already in production and tight with all new stuff).

I think there's going to be a bit move away from all the ruby based stuff (ruby-sass, compass, scss-lint, etc.), so this will be more and more of a 'requirement' for folks to ease the transition.

So, +1 from me!

**Comment by benthemonkey** [comment]
This is definitely something that should be saved for when we majorly restructure the code to perform a single traversal of the AST (see #5)

**Comment by ctumolosus** [comment]
Also, useful if you want people to use this package while `gonzales-pe` is trying to figure out how to solve #321 without having to ignore the whole file. I have the same issue using other tools that rely on it, but they provide a way to disable/enable sections of the source code. I've been waiting for interpolation to be fixed in `gonzales-pe` for a while.

**Comment by DanPurdy** [comment]
I'm afraid this wouldn't solve that problem @ctumolosus. The issue is is that even with this feature `gonzales-pe` would still throw an error on interpolated code even if it was set to be ignored. We can't ignore code without first parsing it to know what it is. `gonzales-pe` parses the whole file and then we act upon what it returns. The way you are looking at this working to ignore interpolation would actually need to be a `gonzales-pe` feature to ignore code as it parses it.

**Comment by natorojr** [comment]
FWIW, I'm another sass-lint user in desperate need of this feature.

Glad to see it's being worked on / considered.

**Comment by mknadler** [comment]
:+1:, currently needing to manually comment out lines with `!important` flags in order to run linter, which is a blocker for integrating it fully into pre-commit hooks + build scripts

**Comment by digitaldonkey** [comment]
I'm waiting this feature desperately too. I can help testing :)

**Comment by gitnik** [comment]
Big +1 for this. Also willing to help with the implementation if steered into the right direction

**Comment by donabrams** [comment]
I'm working on this today, should have a pull by 5pm MST

**Comment by DanPurdy** [comment]
Ha, i've been working on this too, had a few bits working but the amount of tests needed here will be quite extensive and also some of the line numbers seem to cause issues also.

**Comment by bgriffith** [comment]
Get talking guys! We don't want any wasted effort 😛

**Comment by DanPurdy** [comment]
So the way i was approaching it was to have a rule setup similar to the comments rule, it would make a note of which context to ignore and keep a copy of the line number and then before the results are output it would filter through the results object and remove any of the results that matched those from the ignore rule.

It was also taking into account enabling and disabling rules between two blocks such as with scss-lint which worked the same way essentially as above but the different contexts were tough to get exactly right.

**Comment by DanPurdy** [comment]
Happy for you to carry on with this if you can though as I've not had the time recently.

**Comment by donabrams** [comment]
What do you think of https://github.com/donabrams/sass-lint/blob/feature/toggle-rules-in-src/docs/options/toggle-rules-in-src.md and https://github.com/donabrams/sass-lint/blob/feature/toggle-rules-in-src/docs/rules/doc-disabled-rules.md ?

**Comment by DanPurdy** [comment]
I like the extra rule there to enforce the use of documented comments with them. As for the implementation this exactly matches the way scss-lint handles it right?

**Comment by donabrams** [comment]
The approach I have definitely relies on correct line numbers to be reported from rules (I'm sure a bunch of bugs will follow).

I deviated from scss-lint's approach by allowing per-line (like eslint) and making the block syntax more explicit.

Gist of approach: run a rule that returns a mapping of where rules are disabled (both globally and per rule). Filter the results for each "file" via that mapping.

**Comment by DanPurdy** [comment]
Pretty much how I had it yes, I tell you what, lets have a specific feature branch here, I'll create a new branch off of develop for this in the main repo, we should then branch off of that, we can then give this a bit of collaborative testing and tweaking and keep it up to date with develop without directly relying on it. That ok with you?

**Comment by benthemonkey** [comment]
Maybe with this much developer attention on this issue, now is the time to move to the single-traversal approach to our core functionality.

I'm not sure how you could disable certain rules on certain lines right now without modifying all rules to check for comments or having to process all lint warnings after all the rules have run. If we perform a single traversal like eslint, we can simply track which rules are enabled and disabled as we traverse.

**Comment by DanPurdy** [comment]
Ok branch off of this branch `feature/disable-linters`https://github.com/sasstools/sass-lint/tree/feature/disable-linters rather than develop.

**Comment by donabrams** [comment]
I was hoping to have this by next Wednesday (linting cleanup "party" for the team). Obviously not a reason to rush something nasty, but what's the scope here? The only scope I'm worrying about not being able to get done is making sure existing rules return the right line numbers for start/end.

**Comment by donabrams** [comment]
@benthemonkey I'd keep this feature independent of rules. Just work off of results the rules return. Not the most optimal performance, but we def don't want to make writing rules any harder.

**Comment by DanPurdy** [comment]
@donabrams & @benthemonkey agreed, filtering results for now will provide the most straightforward way here. For 2.0 I know @Snugug is keen to refactor all of the rules to allow a more eslint style way of having pluggable rules etc, this will then be able to hook into this disabling linters functionality which will become a core method really.. We should hold back on the single traversal until we reach that point I think as we'll be doubling our work otherwise.

If we submit PR's into that feature branch I set up then we'll be able to work collaboratively on this and everyone will be able to chip in with tests issues and examples while allowing us to move forward with our current release pipeline.

**Comment by donabrams** [comment]
OK, so verifying existing rules is in scope for this branch to be integrated?

**Comment by benthemonkey** [comment]
I honestly don't think the refactor to single-traversal is particularly difficult. Instead of calls to "traverseByType" in each rule, we specify which nodes we'd like to hook into and the core code does one call to "traverse", checking for comments and calling rule files as it goes. I honestly think that it would take as long to move to single traversal AND add disabling linters via comments then to write a version of disabling linters via comments that works with the existing core. Just my thought.

**Comment by DanPurdy** [comment]
If it's crucial to the implementation you've started then yes, if however the line numbers are purely an AST problem then no.

I'm not sure there's issues with line numbers in any rule except indentation which I'm not too fussed about right now while we wait for a stable new version of the AST to release.

@benthemonkey that's a significant change in functionality though, i agree it's probably not a massive change but it's one we don't want to take lightly and would rather we all agree on a way forward for it rather than it get tackled individually we want to make sure our move to pluggable rules and a custom API for distributable rule modules is taken into account.

**Comment by benthemonkey** [comment]
OK, thats sounds fine then. I'll help review your code code and tests. I would be interested to know what the timeline is for eventually getting to single traversal though... @Snugug is there a roadmap to 2.0?

**Comment by DanPurdy** [comment]
Lets open a separate issue up for 2.0 discussion.

**Comment by Snugug** [comment]
I like the approach @donabrams is taking and am confident it will work with both the multi traversal we have now and the future single traversal.

Single pass is something we're waiting to implement for 2.0 and should _not_ be part of this PR. Each PR should solve a single issue.

2.0 roadmap can be tracked in https://github.com/sasstools/sass-lint/issues/400

**Comment by WillsB3** [comment]
What is the status of this issue? Is the temporary enabling and disabling of specific rules only going to be addressed in 2.0 or will this feature land in 1.x at some point in the near future?

**Comment by DanPurdy** [comment]
It's under way in #402 just haven't had the time to get round to it recently

**Comment by WillsB3** [comment]
Great news, thanks for the update. Appreciate all the work that's going into this. Thanks for everyones work on this great tool :+1:

**Comment by donabrams** [comment]
Hey, so finally have time to finish this feature. What needs to be done to close this up?

**Comment by DanPurdy** [comment]
Hey @donabrams There's been some big AST changes so it looks like we just need to go through and check everything again, Sass format should now be a bit more stable with its interpretations of blocks so once that format is working correctly we can review what's been done. I intended to get a bit of extra work done on this recently but as usual time is swallowed up elsewhere.

I merged all of your existing progress into our disable linters branch and opened up #677 into develop to keep the progress in check. So all new PR's should be opened into the disable-linters branch and this will update accordingly.

**Comment by albell** [comment]
+1. Excited to switch over to this linter in the near future but this issue is blocking for our team. Thanks for all the hard work.

**Comment by peduarte** [comment]
Hey guys,
I'm running `sass-lint@1.8.2` and linter toggle via comment doesn't seem to be working. I've gone through the issues and pull requests and I cannot tell whether it's been implemented or still WIP?

Could someone please let me know?

Thanks! 👍

**Comment by DanPurdy** [comment]
It's not been released yet, still WIP in #677 hoping to find some time soon to look into it...

**Comment by peduarte** [comment]
@DanPurdy ok thanks

**Comment by todd-richmond** [comment]
See [http://stackoverflow.com/questions/35136098/have-sass-lint-ignore-a-certain-line] for a thread on this as well

For me, I need to import Google Material Design icons, but the documented css logic causes no-duplicate-properties and no-vendor-prefixes to fire. I therefore want to keep those rules for every other file in my tree except my font.scss file

```css
// http://google.github.io/material-design-icons/
@font-face {
  font-family: 'Material Icons';
  font-style: normal;
  font-weight: 400;
  src: url('~material-design-icons/iconfont/MaterialIcons-Regular.eot'); // IE
  src: local('Material Icons'),
  local('MaterialIcons-Regular'),
  url('~material-design-icons/iconfont/MaterialIcons-Regular.woff2') format('woff2'),
  url('~material-design-icons/iconfont/MaterialIcons-Regular.woff') format('woff'),
  url('~material-design-icons/iconfont/MaterialIcons-Regular.ttf') format('truetype');
}

.material-icons {
  // Firefox
  -moz-osx-font-smoothing: grayscale;
  // WebKit
  -webkit-font-smoothing: antialiased;
  direction: ltr;
   ...
}
```

**Comment by smnbbrv** [comment]
Hi guys, this issue seems to have a long story behind and it does not look like it will be fixed in the closest future.

Is there a way to use a beta / alpha / whatever version? I would be totally fine with that.

**Comment by jackblackCH** [comment]
+1 for this feature!

**Comment by DanPurdy** [comment]
@jackblackCH please see #677 should be available in the next release

@smnbbrv if you want to point your package file at the branch in #677 you can use this functionality right now and we'd appreciate the testing feedback also.
